### PR TITLE
Add script for creating bot user accounts in GitLab via API

### PR DESCRIPTION
**Pull request by fredg02**
Signed-off-by: Frederic Gurr <frederic.gurr@eclipse-foundation.org>

### CODE REVIEW

File: gitlab/createGitLabBotUser.sh [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    pwgen -1 -s -y 24 | pass insert --echo "${pw_store_path}/password"
+  else
+    echo "Found ${gitlab_pass_domain} SSH credentials in password store. Skipping creation..."
+  fi
+}
+
+create_bot_user() {
+  local username=$(pass "${pw_store_path}/username")
+  local email="${username}@eclipse.org"
+  local pw=$(pass "${pw_store_path}/password")
```
Comment (mbarbero) [inline gitlab/createGitLabBotUser.sh:76]:
```suggestion
  local pw="$(pass "${pw_store_path}/password")"
```
https://github.com/koalaman/shellcheck/wiki/SC2046

File: gitlab/createGitLabBotUser.sh [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  if [ "$(curl -s --header "${token_header}" "${api_base_url}/users?username=${username}" | jq .)" == "[]" ]; then
+    echo "Creating bot user ${username}..."
+    add_user_api "${username}" "${email}" "${pw}" "${name}"
+  else
+    echo "User ${username} already exists. Skipping creation..."
+  fi
+}
+
+get_id_from_username() {
+  local username=$(pass "${pw_store_path}/username")
```
Comment (mbarbero) [inline gitlab/createGitLabBotUser.sh:89]:
```suggestion
  local username="$(pass "${pw_store_path}/username")"
```
https://github.com/koalaman/shellcheck/wiki/SC2046

File: gitlab/createGitLabBotUser.sh [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  curl -s --header "${token_header}" "${api_base_url}/users?username=${username}" | jq '.[].id'
+}
+
+add_ssh_key() {
+  # get ID
+  local user_id=$(get_id_from_username)
```
Comment (mbarbero) [inline gitlab/createGitLabBotUser.sh:95]:
```suggestion
  local user_id="$(get_id_from_username)"
```
https://github.com/koalaman/shellcheck/wiki/SC2046

File: gitlab/createGitLabBotUser.sh [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    add_ssh_key_api "${id}" "${username}" "${id_rsa_pub}"
+  else
+    echo "SSH key already exists. Skipping creation..."
+  fi
+}
+
+# create impersonation token
+create_api_token() {
+  # get ID
+  local user_id=$(get_id_from_username)
```
Comment (mbarbero) [inline gitlab/createGitLabBotUser.sh:111]:
```suggestion
  local user_id="$(get_id_from_username)"
```
https://github.com/koalaman/shellcheck/wiki/SC2046

**Review by mbarbero — CHANGES_REQUESTED** [review]
Sorry for the long review. Went quite in length here (minor stuff, but as always with bash, minor stuff can lead to long debugging ;))

File: gitlab/createGitLabBotUser.sh [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+# Bash strict-mode
+set -o errexit
+set -o nounset
+set -o pipefail
+
+IFS=$'\n\t'
+script_name="$(basename ${0})"
+script_folder="$(dirname $(readlink -f "${0}"))"
+
+export password_store_dir=~/.password-store/cbi-pass
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+gitlab_pass_domain="gitlab.eclipse.org"
+
+
+personal_access_token=$(cat ../.localconfig | jq -r '."gitlab-token"')
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+token_header="PRIVATE-TOKEN: ${personal_access_token}"
+api_base_url="https://gitlab.eclipse.org/api/v4"
+
+project_name=${1:-}
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+# verify input
+if [ -z "${project_name}" ]; then
+  printf "ERROR: a project name (e.g. 'technology.cbi' for CBI project) must be given.\n"
+  exit 1
+fi
+
+pw_store_path="cbi-pass/bots/${project_name}/${gitlab_pass_domain}"
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+## GitLab API ##
+
+add_user_api() {
+  local username=$1
+  local email=$2
+  local pw=$3
+  local name=$4 #display name
+
+  curl -s --header "${token_header}" --request POST "${api_base_url}/users" --data "username=${username}" --data "email=${email}" --data "password=${pw}" --data "name=${name}" | jq .
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+add_ssh_key_api() {
+  local id=$1
+  local title=$2
+  local key=$3
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+  curl -s --header "${token_header}" --request POST "${api_base_url}/users/${id}/keys" --data-urlencode "title=${title}" --data-urlencode "key=${key}" | jq .
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+
+create_credentials_in_pass() {
+  if [[ ! -f "${password_store_dir}/bots/${project_name}/${gitlab_pass_domain}/id_rsa.gpg" ]]; then
+    echo "Creating GitLab SSH credentials in SSH..."
+    pushd "${script_folder}/.."
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    popd
+    # create password
+    pwgen -1 -s -y 24 | pass insert --echo "${pw_store_path}/password"
+  else
+    echo "Found ${gitlab_pass_domain} SSH credentials in password store. Skipping creation..."
+  fi
+}
+
+create_bot_user() {
+  local username=$(pass "${pw_store_path}/username")
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh (hunk 1 above)
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh (hunk 2 above)
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    echo "Creating bot user ${username}..."
+    add_user_api "${username}" "${email}" "${pw}" "${name}"
+  else
+    echo "User ${username} already exists. Skipping creation..."
+  fi
+}
+
+get_id_from_username() {
+  local username=$(pass "${pw_store_path}/username")
+  curl -s --header "${token_header}" "${api_base_url}/users?username=${username}" | jq '.[].id'
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh (hunk 3 above)
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+add_ssh_key() {
+  # get ID
+  local user_id=$(get_id_from_username)
+
+  # if SSH key already exists, skip
+  if [ "$(curl -s --header "${token_header}" "${api_base_url}/users/${user_id}/keys" | jq .)" == "[]" ]; then
+    echo "Creating SSH key for ${username}..."
+    # read ssh public key from pass
+    id_rsa_pub=$(pass "${pw_store_path}/id_rsa.pub")
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh (hunk 4 above)
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    echo "SSH key already exists. Skipping creation..."
+  fi
+}
+
+# create impersonation token
+create_api_token() {
+  # get ID
+  local user_id=$(get_id_from_username)
+
+  name="CI token"
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+create_bot_user() {
+  local username=$(pass "${pw_store_path}/username")
+  local email="${username}@eclipse.org"
+  local pw=$(pass "${pw_store_path}/password")
+  local name="${short_name} bot user"
+
+   # if bot user already exists, skip
+  if [ "$(curl -s --header "${token_header}" "${api_base_url}/users?username=${username}" | jq .)" == "[]" ]; then
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: > What about defining a convention about variables?

:+1:  Sounds like a good plan.

> You can also use jq -e which exit with an error code for empty arrays or null:

I had to change the logic around to make this work.

File: gitlab/createGitLabBotUser.sh [hunk 18]
```diff
@@ -15,112 +15,124 @@ set -o nounset
...
+PASSWORD_STORE_DIR="$(jq -r '.["password-store"]["cbi-dir"]' "${SCRIPT_FOLDER}/../.localconfig")"
+PASSWORD_STORE_DIR="$(readlink -f "${PASSWORD_STORE_DIR/#~\//${HOME}/}")"
+export PASSWORD_STORE_DIR
 
+GITLAB_PASS_DOMAIN="gitlab.eclipse.org"
 
-personal_access_token=$(cat ../.localconfig | jq -r '."gitlab-token"')
-token_header="PRIVATE-TOKEN: ${personal_access_token}"
-api_base_url="https://gitlab.eclipse.org/api/v4"
+PERSONAL_ACCESS_TOKEN="$(cat ../.localconfig | jq -r '."gitlab-token"')"
```
Comment (mbarbero) [inline gitlab/createGitLabBotUser.sh:32]: Should be `${SCRIPT_FOLDER}/../.localconfig`

File: gitlab/createGitLabBotUser.sh [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+script_name="$(basename ${0})"
+script_folder="$(dirname $(readlink -f "${0}"))"
+
+export password_store_dir=~/.password-store/cbi-pass
+gitlab_pass_domain="gitlab.eclipse.org"
+
+
+personal_access_token=$(cat ../.localconfig | jq -r '."gitlab-token"')
+token_header="PRIVATE-TOKEN: ${personal_access_token}"
+api_base_url="https://gitlab.eclipse.org/api/v4"
```
[the remark this answers is not in the document]
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: I'd rather change a global variable in a script than exporting a environment variable, but that's a personal preference and I won't fight for it. I can't promise to remember this best practice in the next bash script though. :stuck_out_tongue:

File: gitlab/createGitLabBotUser.sh [hunk 20]
```diff
@@ -15,112 +15,124 @@ set -o nounset
...
 
-  curl -s --header "${token_header}" --request POST "${api_base_url}/users" --data "username=${username}" --data "email=${email}" --data "password=${pw}" --data "name=${name}" | jq .
+  curl -sSL --header "${TOKEN_HEADER}" --request POST "${API_BASE_URL}/users" --data "username=${username}" --data "password=${pw}" --data "email=${email}" --data "name=${name}"
 }
 
 add_ssh_key_api() {
-  local id=$1
-  local title=$2
-  local key=$3
+  local id"=$1"
```
Comment (mbarbero) [inline gitlab/createGitLabBotUser.sh:59]: Compile error: id="$1"

File: gitlab/createGitLabBotUser.sh [hunk 21]
```diff
@@ -15,112 +15,124 @@ set -o nounset
...
-  curl -s --header "${token_header}" "${api_base_url}/users?username=${username}" | jq '.[].id'
+  local username="$1"
+  curl -s --header "${TOKEN_HEADER}" "${API_BASE_URL}/users?username=${username}" | jq -r '.[].id'
 }
 
 add_ssh_key() {
   # get ID
-  local user_id=$(get_id_from_username)
+  local username="$1"
+  local user_id="$(get_id_from_username ${username})"
```
Comment (mbarbero) [inline gitlab/createGitLabBotUser.sh:102]: don't forget quotes inside $() neither: `get_id_from_username "${username}"`

File: gitlab/createGitLabBotUser.sh [hunk 22]
```diff
@@ -15,112 +15,124 @@ set -o nounset
...
+    add_ssh_key_api "${user_id}" "${username}" "${id_rsa_pub}" | jq .
   fi
 }
 
 # create impersonation token
 create_api_token() {
   # get ID
-  local user_id=$(get_id_from_username)
+  local username="$1"
+  local user_id="$(get_id_from_username ${username})"
```
Comment (mbarbero) [inline gitlab/createGitLabBotUser.sh:120]: "${username}"

**Review by mbarbero — CHANGES_REQUESTED** [review]

File: gitlab/createGitLabBotUser.sh (hunk 18 above)
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh (hunk 20 above)
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: :facepalm:

File: gitlab/createGitLabBotUser.sh (hunk 21 above)
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh (hunk 22 above)
Comment (fredg02) [reply gitlab/createGitLabBotUser.sh]: Fixed

File: gitlab/createGitLabBotUser.sh (hunk 20 above)
Comment (mbarbero) [reply gitlab/createGitLabBotUser.sh]: review FTW ;)

**Review by mbarbero — APPROVED** [review]

**Pull request merged**